### DISCUSSION TITLE
Session not working well

### DISCUSSION THREAD

**Discussion by maki10** [laravel/framework#38978] [Q&A]
- Laravel Version: 6.20.34 | 8.11.2
- PHP Version: 7.3.10 | ?
- Database Driver & Version: 5.8 | ?

### Description:
I have an issue with session drivers "file" and "redis". My app is a simple e-store, but in order to work well with a third-party vendor, I must use several sessions in one checkout, for example 20+. My issue starts when using too many sessions.

### Steps To Reproduce:
If you go to: https://laravelplayground.com/#/snippets/bdbd0f74-e871-4a9b-911c-88a4c8eca9cb and uncomment just one line in CheckoutService.php (From line 24 until line 44) session will start to fails.

**Comment by X-Coder264** [comment]
You cannot call the session in the constructor like that. The reason being that your `CheckoutService` is being constructed in `CheckoutController` constructor which gets created before the `\Illuminate\Session\Middleware\StartSession` middleware gets run so the session hasn't been started in that point in time.

The controller gets constructed in [\Illuminate\Routing\Route::gatherMiddleware](https://github.com/laravel/framework/blob/v8.61.0/src/Illuminate/Routing/Route.php#L1001) as Laravel has a weird feature (that should IMO be removed) that allows you to add middleware in the controller constructor which leads to problems like this one (as the controller gets created before any of the route middleware gets executed).

**Comment by maki10** [comment]
@X-Coder264 Yes, maybe I not provided a good example. Even if I move sessions "setters" from `__construct` and place in the first called method, an issue still persists.. Take a look at https://laravelplayground.com/#/snippets/b119ef2c-4269-477d-8d09-544332df71d3

**Comment by maki10** [comment]
Hey, @driesvints I saw that you are close my issue and mark it like Q&A, but I am sure that this is a bug... I am not sure why this is happening but will take a deeper look soon and find what causes it.
